### DISCUSSION TITLE
Why does DataGrid require a style added to App.axaml ?

### DISCUSSION THREAD

**Discussion by moon6969** [AvaloniaUI/Avalonia#6247] [General]
I spent hours trying to figure out why my Datagrid had no errors or warnings, but would not appear in the window. I assumed it was a binding error or issue with my ViewModel.

I did reference https://docs.avaloniaui.net/docs/controls/datagrid, but initially ignored the "Add required styles to App.axaml" section since I was just hacking up a quick proof of concept and didn't care what it looked like... even in prod development I would typically not bother with styling until everything was wired up.

I don't normally bother with the visual designer - but in desperation I checked it, and it was only there that I noticed...
`Unable to resolve type DataGrid from namespace https://github.com/avaloniaui`
Why is there no warning when running the app?

Is there any documentation/discussion on _why_ DataGrid requires it's own style included?

I would like to update [DataGrid documentation](https://docs.avaloniaui.net/docs/controls/datagrid) with something like:
_⚠️ Control will not be visible without the DataGrid style (see [link to explanation])_

What other controls require their own style?

**Comment by maxkatz6** [comment]
> I would like to update DataGrid documentation with something like:
⚠️ Control will not be visible without the DataGrid style (see [link to explanation])

It says already, that styles **are required**.
The only one real problem of current documentation, that Fluent theme was not mentioned there.
I updated the page and solved that problem. Thank you for bringing some attention.

> What other controls require their own style?

All of them. No exceptions at this moment.
When you create application from the template, build-in theme with styles is already included there (Default theme for older Avalonia versions and Fluent theme for 0.10.0 and newer). So, it still requires style include, but it's done as part of default template.
These themes cannot contain styles for controls outside of Avalonia.Control or, in other words, controls that are not available in Avalonia.Controls package.

DataGrid is not included with the template, and it's distributed with separated package (although Avalonia.Diagnostics might reference it transitively). That's why it requires other styles to include.

Also, for the same reason any third-party control will also require the same.

**Reply by maxkatz6** [reply]
> All of them. No exceptions at this moment.

By this I mean that in the future we might support implicit style include, if library has Generic.xaml created (something like [WPF has](https://stackoverflow.com/questions/1228875/what-is-so-special-about-generic-xaml), couldn't found proper docs page). With this feature it will be enough to just reference nuget package.

But it's not on the roadmap for nearest releases. Although if anybody willing to help to implement it, they are always welcomed.

**Reply by moon6969** [reply]
Thanks for reply.

I just installed Avalonia & VS extension and started using the supplied templates - I don't recall installing any separate DataGrid package, and I certainly didn't nuget Datagrid in my test project. I can see the Diagnostics package though.

As new user, it certainly wasn't clear to me that styles are required just to get the controls to appear at all - my naive understanding was that controls know how to display themselves, and that styling just changes the appearance. A quick revist just now of Avalonia's "Getting Started" and "WPF Developer Tips" pages doesn't really indicate otherwise to me either.

Even looking at the DataGrid in the runtime F12 DevTools I couldn't see that the issue was a missing style file.

> It says already, that styles **are required**.

It would helpful for newer users to explain the consequence of not including it, or perhaps include a tip "if your control is not visible, check you have included the style".
